### PR TITLE
[SPARK-30320][SQL] Fix insert overwrite to DataSource table with dynamic partition error

### PR DESCRIPTION
**Pull request by WinkerDu**
### What changes were proposed in this pull request?
This PR fix insert overwrite table error when running multiple speculative task. This PR introduces task attempt id appended to dynamic partition staging dir to identify different speculative task output dir rather than all speculative tasks use same output dir. CommitTask in HadoopMapReduceCommitProtocol carries attempt id in TaskCommitMessage

### Why are the changes needed?
This PR fix insert overwrite to DataSource table with dynamic partition error when running multiple task attempts. Suppose there are one task attempt and one speculative task attempt, the speculative would raise FileAlreadyExistsException because of same staging dir attempt tasks commit

### How was this patch tested?
Added UT

### CODE REVIEW

**Comment by WinkerDu** [comment]
@xuanyuanking @LuciferYang @LinhongLiu pls have a review

**Comment by WinkerDu** [comment]
cc @cloud-fan

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala [hunk 1]
```diff
@@ -167,8 +169,8 @@ class HadoopMapReduceCommitProtocol(
     committer.commitJob(jobContext)
 
     if (hasValidPath) {
-      val (allAbsPathFiles, allPartitionPaths) =
-        taskCommits.map(_.obj.asInstanceOf[(Map[String, String], Set[String])]).unzip
+      val (allAbsPathFiles, allPartitionPaths, successAttemptIDs) =
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala]: Should we add new implicit function like '
```
implicit def asPair(x:(Map[String, String], Set[String], String)) = (x._1, (x._2, x._3))
```
before line 172, then we can `unzip` taskCommits as `(Map[String, String], (Set[String], String))` and reduce re-`zip` at line 188

Comment (WinkerDu) [reply core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala]: Sounds reasonable, I'll try implicit conversion

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 2]
```diff
@@ -1521,4 +1521,10 @@ package object config {
     .bytesConf(ByteUnit.BYTE)
     .createOptional
 
+  private[spark] val MAX_LOCAL_TASK_FAILURES = ConfigBuilder("spark.task.local.maxFailures")
+    .doc("The max failure times for a task while SparkContext running in Local mode, " +
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: How could you launch speculative task when running under local mode?

Comment (WinkerDu) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: In UT class InsertWithMultipleTaskAttemptSuite, I don't expect launching speculative task in local mode. Actually, I make a customized commit protocol named "InsertExceptionCommitProtocol" in InsertWithMultipleTaskAttemptSuite, which overriding commitTask method to fail the first commit task on purpose and restore to normal in subsequent commit tasks. This scene is similar to what happened in speculative tasks with failure.

**Comment by WinkerDu** [comment]
cc @dongjoon-hyun  please have a review?

File: core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala [hunk 3]
```diff
@@ -183,22 +187,30 @@ class HadoopMapReduceCommitProtocol(
...
+                // need to double check if finalPartPath exists before rename.
+                // Also in our case, based on the spec, delete returns false only when finalPartPath
+                // does not exist. When this happens, we need to take action if parent of
+                // finalPartPath also does not exist(e.g. the scenario described on SPARK-23815),
+                // because FileSystem API spec on rename op says the rename dest(finalPartPath)
+                // must have a parent that exists, otherwise we may get unexpected result
+                // on the rename.
+                fs.mkdirs(finalPartPath.getParent)
+              }
+              fs.rename(new Path(s"$stagingDir/$successAttemptID", part), finalPartPath)
```
Comment (koertkuipers) [inline core/src/main/scala/org/apache/spark/internal/io/HadoopMapReduceCommitProtocol.scala:206]: do i understand it correctly that part here is a directory (e.g. x=1/y=2), not a file? so a directory full of files is being moved.
if so couldn't multiple tasks write to the same partition? and then wouldnt these moves conflict with each other?

**Comment by ramesh-muthusamy** [comment]
@WinkerDu  I think the same issue is being worked upon in the PR for the ticket. https://issues.apache.org/jira/browse/SPARK-27194
https://github.com/apache/spark/pull/26339
Please validate

**Pull request closed without merging**